libvast/src/system/index.cpp

```diff
@@ -214,212 +193,346 @@ caf::actor index_state::next_worker() {
   // num_partitions partitions or run out of candidates.
-  {
-    auto i = lookup.partitions.begin();
-    auto last = lookup.partitions.end();
-    for (; i != last && result.size() < num_partitions; ++i)
-      spin_up(*i);
-    lookup.partitions.erase(lookup.partitions.begin(), i);
+  auto it = lookup.partitions.begin();
+  auto last = lookup.partitions.end();
+  while (it != last && result.size() < num_partitions) {
+    auto partition_id = *it++;
+    if (auto partition_actor = spin_up(partition_id))
+      result.push_back(std::make_pair(partition_id, partition_actor));
   }
+  lookup.partitions.erase(lookup.partitions.begin(), it);
+  VAST_DEBUG(self, "launched", result.size(),
+             "await handlers to fill the pending query map");
   return result;
 }
 
-query_map
-index_state::launch_evaluators(pending_query_map pqm, expression expr) {
-  query_map result;
-  for (auto& [id, eval] : pqm) {
-    std::vector<caf::actor> xs{self->spawn(evaluator, expr, std::move(eval))};
-    result.emplace(id, std::move(xs));
+/// Sends an expression to all partition actors passed into this function,
+/// and collects the resulting evaluators.
+/// @param c Continuation that takes a single argument of type
+/// `caf::expected<query_map>`. The continuation will be called in the
+/// actor context of `self`.
+//
+// TODO: At some point we should add some more template magic on top of
+// this and turn it into a generic functions that maps
+//
+//   (map from U to A, request param pack R, result handler with param X) ->
+//   expected<map from U to X>
+template <typename Continuation>
+void await_evaluators_then(
+  caf::stateful_actor<index_state>* self, const expression& expr,
+  const std::vector<std::pair<vast::uuid, caf::actor>>& actors,
+  Continuation c) {
+  struct counter {
+    size_t received;
+    query_map qm;
+  };
+  auto expected = actors.size();
+  auto shared_counter = std::make_shared<counter>();
+  for (auto& [id, actor] : actors) {
+    auto partition_id = id; // Can't use structured binding inside lambda.
+    self->request(actor, caf::infinite, expr)
+      .then(
+        [=](caf::actor evaluator) {
+          auto received = ++shared_counter->received;
+          if (evaluator)
+            shared_counter->qm.emplace(partition_id, std::vector<caf::actor>{evaluator});
+          if (received == expected)
+            c(std::move(shared_counter->qm));
+        },
+        [=](caf::error err) {
+          // Don't increase `received` to ensure the sucess handler never gets
+          // called.
+          c(err);
+        });
   }
-  return result;
 }
 
-void index_state::add_flush_listener(caf::actor listener) {
-  VAST_DEBUG(self, "adds a new 'flush' subscriber:", listener);
-  flush_listeners.emplace_back(std::move(listener));
-  detail::notify_listeners_if_clean(*this, *stage);
+path index_state::index_filename(path basename) const {
+  return basename / dir / "index.bin";
```

Comment:
We probably don't need `basename` if we do all I/O through the filesystem actor.

Reply:
Not sure. I added the basename here without thinking about it too much and removed the TODO. Don't know why that TODO existed in the first place.